From 0320ea65de3bfc694d6f24bc410433a544555292 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Thu, 29 Jul 2021 12:28:20 +1000
Subject: [PATCH] wined3d: Merged ATOMIC_IMIN and ATOMIC_UMIN together

Batman: Arkham Knight.

0670:fixme:d3d_shader:shader_glsl_atomic Unhandled opcode 0x6 for unsigned integers.
0670:fixme:d3d_shader:shader_glsl_atomic Unhandled opcode 0x7 for unsigned integers.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=52233
---
 dlls/wined3d/glsl_shader.c | 36 ++++++++----------------------------
 1 file changed, 8 insertions(+), 28 deletions(-)

diff --git a/dlls/wined3d/glsl_shader.c b/dlls/wined3d/glsl_shader.c
index 5cccd31207b..9928134c799 100644
--- a/dlls/wined3d/glsl_shader.c
+++ b/dlls/wined3d/glsl_shader.c
@@ -5628,26 +5628,30 @@ static void shader_glsl_atomic(const struct wined3d_shader_instruction *ins)
                 op = "imageAtomicAdd";
             break;
         case WINED3DSIH_ATOMIC_IMAX:
+        case WINED3DSIH_ATOMIC_UMAX:
         case WINED3DSIH_IMM_ATOMIC_IMAX:
+        case WINED3DSIH_IMM_ATOMIC_UMAX:
             if (is_tgsm)
                 op = "atomicMax";
             else
                 op = "imageAtomicMax";
-            if (data_type != WINED3D_DATA_INT)
+            if (data_type != WINED3D_DATA_INT && data_type != WINED3D_DATA_UINT)
             {
-                FIXME("Unhandled opcode %#x for unsigned integers.\n", ins->handler_idx);
+                FIXME("Unhandled opcode %#x for integers.\n", ins->handler_idx);
                 return;
             }
             break;
         case WINED3DSIH_ATOMIC_IMIN:
+        case WINED3DSIH_ATOMIC_UMIN:
         case WINED3DSIH_IMM_ATOMIC_IMIN:
+        case WINED3DSIH_IMM_ATOMIC_UMIN:
             if (is_tgsm)
                 op = "atomicMin";
             else
                 op = "imageAtomicMin";
-            if (data_type != WINED3D_DATA_INT)
+            if (data_type != WINED3D_DATA_INT && data_type != WINED3D_DATA_UINT)
             {
-                FIXME("Unhandled opcode %#x for unsigned integers.\n", ins->handler_idx);
+                FIXME("Unhandled opcode %#x for integers.\n", ins->handler_idx);
                 return;
             }
             break;
@@ -5658,30 +5662,6 @@ static void shader_glsl_atomic(const struct wined3d_shader_instruction *ins)
             else
                 op = "imageAtomicOr";
             break;
-        case WINED3DSIH_ATOMIC_UMAX:
-        case WINED3DSIH_IMM_ATOMIC_UMAX:
-            if (is_tgsm)
-                op = "atomicMax";
-            else
-                op = "imageAtomicMax";
-            if (data_type != WINED3D_DATA_UINT)
-            {
-                FIXME("Unhandled opcode %#x for signed integers.\n", ins->handler_idx);
-                return;
-            }
-            break;
-        case WINED3DSIH_ATOMIC_UMIN:
-        case WINED3DSIH_IMM_ATOMIC_UMIN:
-            if (is_tgsm)
-                op = "atomicMin";
-            else
-                op = "imageAtomicMin";
-            if (data_type != WINED3D_DATA_UINT)
-            {
-                FIXME("Unhandled opcode %#x for signed integers.\n", ins->handler_idx);
-                return;
-            }
-            break;
         case WINED3DSIH_ATOMIC_XOR:
         case WINED3DSIH_IMM_ATOMIC_XOR:
             if (is_tgsm)
-- 
2.30.2

